semantics-3.0/Semantics.hs

```diff
@@ -236,206 +227,219 @@ data ReduceWarning = ReduceNoWarning
-        | endSlot < timeout -> NotReduced
+        in if endSlot < timeout then NotReduced
         -- if timeout in the past – reduce to timeout continuation
-        | timeout <= startSlot -> Reduced ReduceNoWarning ReduceNoEffect state cont
+        else if timeout <= startSlot then Reduced ReduceNoWarning Nothing state cont
         -- if timeout in the slot range – issue an ambiguity error
-        | otherwise -> ReduceError ReduceAmbiguousSlotInterval
-      where
-        startSlot = fst $ slotInterval env
-        endSlot   = snd $ slotInterval env
-    Let valId val cont -> Reduced warn ReduceNoEffect newState cont
-      where
-        evVal = evalValue env state val
+        else AmbiguousSlotIntervalError
+
+    Let valId val cont -> let
+        evaluatedValue = evalValue env state val
         boundVals = boundValues state
-        newState = state { boundValues = Map.insert valId evVal boundVals }
+        newState = state { boundValues = Map.insert valId evaluatedValue boundVals }
         warn = case Map.lookup valId boundVals of
-              Just oldVal -> ReduceShadowing valId oldVal evVal
+              Just oldVal -> ReduceShadowing valId oldVal evaluatedValue
               Nothing -> ReduceNoWarning
+        in Reduced warn Nothing newState cont
 
 
--- REDUCE ALL
-
-data ReduceAllResult = ReducedAll [ReduceWarning] [ReduceEffect] State Contract
-                     | ReduceAllError ReduceError
-  deriving (Eq,Ord,Show,Read)
+data ReduceAllResult = ReduceAllSuccess [ReduceWarning] [Payment] State Contract
+                     | ReduceAllAmbiguousSlotIntervalError
+  deriving (Eq,Ord,Show)
 
--- Reduce until it cannot be reduced more
-reduceAllAux
-  :: Environment -> State -> Contract -> [ReduceWarning] -> [ReduceEffect] -> ReduceAllResult
-reduceAllAux env state contract warnings effects = case reduce env state contract of
-    Reduced warning effect newState cont -> let
-        newWarnings = if warning == ReduceNoWarning then warnings else warning : warnings
-        newEffects  = if effect  == ReduceNoEffect  then effects  else effect : effects
-        in  reduceAllAux env newState cont newWarnings newEffects
-    ReduceError err -> ReduceAllError err
-    NotReduced -> ReducedAll (reverse warnings) (reverse effects) state contract
+-- | Reduce a contract until it cannot be reduced more
+reduceContractUntilQuiescent :: Environment -> State -> Contract -> ReduceAllResult
+reduceContractUntilQuiescent env state contract = let
+    reduceAll
+      :: Environment -> State -> Contract -> [ReduceWarning] -> [Payment] -> ReduceAllResult
+    reduceAll env state contract warnings effects =
+        case reduceContractStep env state contract of
+            Reduced warning effect newState cont -> let
 
+                newWarnings = if warning == ReduceNoWarning then warnings
+                              else warning : warnings
+                newEffects  = case effect of
+                    Just eff -> eff : effects
```

Comment:
Maybe use Data.Maybe.maybe here.

Reply:
That was the first thing I did. But I changed it towards more explicit code. Maybe.maybe saves us 2 lines of code but requires a reader to know what it does.
Anyway, I'm about to resurrect the ReduceEffect data type, so it doesn't matter :)

Reply:
I have to admit I can never remember what `Maybe.maybe` does if I haven't used it for a while.

Reply:
I think you are right, I had to check the documentation :)